R/get_source_expressions.R

```diff
@@ -592,88 +592,78 @@ tab_offsets <- function(tab_columns) {
 # This function wraps equal assign expressions in a parent expression so they
 # are the same as the corresponding <- expression
 fix_eq_assigns <- function(pc) {
-  if (is.null(pc)) {
-    return(NULL)
+  if (is.null(pc) || any(c("equal_assign", "expr_or_assign_or_help") %in% pc$token)) {
+    return(pc)
   }
 
   eq_assign_locs <- which(pc$token == "EQ_ASSIGN")
-  if (length(eq_assign_locs) == 0L ||
-    nrow(pc) %in% eq_assign_locs || # check whether the equal-assignment is the final entry
-    any(c("equal_assign", "expr_or_assign_or_help") %in% pc$token)) {
+  # check whether the equal-assignment is the final entry
+  if (length(eq_assign_locs) == 0L || utils::tail(eq_assign_locs, 1L) == nrow(pc)) {
     return(pc)
   }
 
   prev_locs <- vapply(eq_assign_locs, prev_with_parent, pc = pc, integer(1L))
   next_locs <- vapply(eq_assign_locs, next_with_parent, pc = pc, integer(1L))
-  expr_locs <- (function(x) {
-    x[is.na(x)] <- FALSE
-    !x
-    })(prev_locs == lag(next_locs))
+  expr_locs <- prev_locs != lag(next_locs)
+  expr_locs[is.na(expr_locs)] <- TRUE
 
   id_itr <- max(pc$id)
 
-  n_expr <- sum(expr_locs)
-
-  line1 <- integer(n_expr)
-  col1 <- integer(n_expr)
-
-  line2 <- integer(n_expr)
-  col2 <- integer(n_expr)
-
-  id <- integer(n_expr)
-
-  parent <- integer(n_expr)
-
-  token <- character(n_expr)
-
-  terminal <- logical(n_expr)
-
-  text <- character(n_expr)
+  true_locs <- which(expr_locs)
+  n_expr <- length(true_locs)
+
+  supplemental_content <- data.frame(
+    line1 = integer(n_expr),
+    col1 = integer(n_expr),
+    line2 = integer(n_expr),
+    col2 = integer(n_expr),
+    id = integer(n_expr),
+    parent = integer(n_expr),
+    token = character(n_expr),
+    terminal = logical(n_expr),
+    text = character(n_expr),
+    stringsAsFactors = FALSE
+  )
 
-  true_locs <- which(expr_locs == TRUE)
-  for (i in seq_along(true_locs)) {
+  for (i in seq_len(n_expr)) {
     start <- true_locs[i]
 
     end <- true_locs[i]
     j <- end + 1L
-    while (j <= length(expr_locs) && expr_locs[j] == FALSE) {
```

Comment:
this can likely be vectorized (looks like a `findInterval()` thing?) but it's a bit hard to test as my 3.6.3 doesn't run this region except for a trivial error test case (`{ x = }` in test-error.R)

Reply:
are we confident our tests are sufficient? don't want to touch it unless we know it's covered well (i.e., in a case where `n_expr > 1`)

Reply:
The line isn't hit on R 4.2.0 either...

Reply:
covr should tell us that, and i think it's likely hit (it's hit in the weird edge case of syntax error `{ x = }` in test-error.R), but the hit is in a trivial case when `n_expr=1`, so we don't really learn anything.

what'll be interesting is to stop() if `n_expr > 1`... basically wrote as such in the TODO.